src/apps/EDGE3D.cpp

```diff
@@ -0,0 +1,125 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "EDGE3D.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#include "AppsData.hpp"
+#include "common/DataUtils.hpp"
+
+#include <cmath>
+
+
+namespace rajaperf
+{
+namespace apps
+{
+
+
+EDGE3D::EDGE3D(const RunParams& params)
+  : KernelBase(rajaperf::Apps_EDGE3D, params)
+{
+  setDefaultProblemSize(100*100*100);  // See rzmax in ADomain struct
+  setDefaultReps(10);
+  constexpr int num_quadrature_points = NQ_1D*NQ_1D*NQ_1D;
+
+  Index_type rzmax = std::cbrt(getTargetProblemSize())+1;
+  m_domain = new ADomain(rzmax, /* ndims = */ 3);
+
+  m_array_length = m_domain->nnalls;
+  int number_of_elements = m_domain->lpz+1 - m_domain->fpz;
+
+  setActualProblemSize( number_of_elements );
+
+  setItsPerRep( number_of_elements );
+  setKernelsPerRep(1);
+
+  constexpr int matrix_size = NB*NB;
+  constexpr int basis_size = NB;
+
+  constexpr int reals_per_element =
+    2*NQ_1D + // quadrature size and weights
+    12*basis_size + // 3 basis, 3 tbasis, 3 dbasis, 3 tdbasis
+    matrix_size;
+
+  // touched data size, not actual number of stores and loads ?
+  setBytesPerRep( number_of_elements*reals_per_element*sizeof(Real_type) );
+
+  // Only consider the operations in the innermost loop
+  // these are done for each element of a matrix of size matrix_size
+  // and this matrix is computed num_quadrature_points times
+  constexpr int innermost_flops =
+    6 + // detjwgts*(txm*txp + tym*typ + tzm*tzp)
+    6 + // detjwgts*(dtxm*dtxp + dtym*dtyp + dtzm*dtzp)
+    1;  // x = Mtemp + Stemp
+
+  constexpr int flops_per_element = num_quadrature_points*matrix_size*innermost_flops;
+
+  setFLOPsPerRep(number_of_elements * flops_per_element);
+
+  // No idea what this is?
+  checksum_scale_factor = 0.001 *
```

Comment:
I believe this is used to scale terms in the output array in a manner that helps avoid getting the right answer by accident. I'm forgetting the details, but perhaps @rhornung67  can help explain a bit.

Reply:
Depending on a kernel's properties (default problem size, default # samples, how a checksum is calculated, typical magnitude of result values used to calculate checksums, etc.) we sometimes ad a scale factor to the checksum calculation to avoid overflow and accuracy issues. Typically, we try to shoot for checksum differences < 1e-8 between variants for correct results.